README.md
# SearchEngine

### 目录结构

- **./data**：抓取的数据
- **./index**: lucene建立的索引
- **./lucene-4.0.0**: lucene .jar包
- **./src**: java源码
- **netty-all-4.1.11.Final.jar**：netty网络库

### 运行方法

1. 用最新的Idea打开项目
2. 运行SingerDicBuilder。 **./src**目录中会生成**singer.dic**文件，包含了所有歌手名。
3. 运行IndexBuilder。 **./index**目录中会生成**lucene索引**
4. 运行Server。服务器开始运行，可以通过网络连接搜索了。

### 查询方法

首先用 `nc -Cv localhost 8080`命令与服务器建立连接，然后查询：

```
all: 彩虹
```

返回：

```
hits: 10

id: 26402184
song: *彩虹*
singer: '李霄云'

id: 156410
song: *彩虹*
singer: '伍佰'

id: 28253225
song: *彩虹*
singer: '张艺兴','吴亦凡'

id: 326887
song: *彩虹*
singer: '张惠妹'

id: 185809
song: *彩虹*
singer: '周杰伦'
...
```

周杰伦的彩虹不在第一名，这时候点击周杰伦的彩虹(id + 点击次数)：

```
update：185809 1
all：彩虹
```

返回

```
hits: 10

id: 185809
song: *彩虹*
singer: '周杰伦'

id: 26402184
song: *彩虹*
singer: '李霄云'

id: 156410
song: *彩虹*
singer: '伍佰'
...
```

查询：

```
signer: 薛之谦
lrc: 该配合你演出的我
(空行)
```

返回：

```
hits: 2
(空行)
id: 32507038
song: 演员
singer: '*薛之谦*'
lrc: :02.220]*该**配合**你**演出*的*我*演视而不见
[01
(空行)
id: 443277013
song: 火星人来过
singer: '*薛之谦*'
lrc: :02.070]请你*配合**我*  一起难过
(空行)
```

查询：

```
song:love
(空行)
```

返回

```
song:love
hits: 100
(空行)
id: 5246556
song: Is this *love*
singer: '逃跑计划'
(空行)
id: 26211276
song: This Is *Love*
singer: 'will.i.am','Eva Simons'
(空行)
id: 25638055
song: For *Love*
singer: '陈冠希'
(空行)
id: 25706276
song: Is This *Love*
singer: '逃跑计划'
(空行)
id: 306653
song: *Love*
singer: '徐佳莹'
(空行)
id: 460476155
song: *Love*
singer: 'Lana Del Rey'
(空行)
id: 1297705
song: No *Love*
singer: 'Lil Wayne','Eminem'
...
```

点击**徐佳莹的Love**(id + 点击次数)：

```
update: 306653 1
(空行)
```

返回:

```
hits: 0
(空行)
```

再次查询:

```
song:love
(空行)
```

返回:

```
id: 306653
song: *Love*
singer: '徐佳莹'
(空行)
id: 5246556
song: Is this *love*
singer: '逃跑计划'
(空行)
id: 26211276
song: This Is *Love*
singer: 'will.i.am','Eva Simons'
(空行)
id: 25638055
song: For *Love*
singer: '陈冠希'
...
```

### 注意

- 网络通信中所有的换行符均为“**\r\n**”
- highlight部分用“*****”标记
- 项目依赖的jar包：

![](jar.png)

lucene-4.0.0/CHANGES.txt
Lucene Change Log

For more information on past and future Lucene versions, please see:
http://s.apache.org/luceneversions

For "contrib" changes prior to 4.0, please see:
http://svn.apache.org/repos/asf/lucene/dev/tags/lucene_solr_3_6_0/lucene/contrib/CHANGES.txt

======================= Lucene 4.0.0 =======================

Changes in backwards compatibility policy

* LUCENE-4392: Class org.apache.lucene.util.SortedVIntList has been removed.
  (Adrien Grand)

* LUCENE-4393: RollingCharBuffer has been moved to the o.a.l.analysis.util
  package of lucene-analysis-common. (Adrien Grand)

New Features

* LUCENE-1888: Added the option to store payloads in the term
  vectors (IndexableFieldType.storeTermVectorPayloads()). Note 
  that you must store term vector positions to store payloads.
  (Robert Muir)

* LUCENE-3892: Add a new BlockPostingsFormat that bulk-encodes docs,
  freqs and positions in large (size 128) packed-int blocks for faster
  search performance.  This was from Han Jiang's 2012 Google Summer of
  Code project (Han Jiang, Adrien Grand, Robert Muir, Mike McCandless)

* LUCENE-4323: Added support for an absolute maximum CFS segment size
  (in MiB) to LogMergePolicy and TieredMergePolicy.
  (Alexey Lef via Uwe Schindler)
  
* LUCENE-4339: Allow deletes against 3.x segments for easier upgrading.
  Lucene3x Codec is still otherwise read-only, you should not set it
  as the default Codec on IndexWriter, because it cannot write new segments.
  (Mike McCandless, Robert Muir)

* SOLR-3441: ElisionFilterFactory is now MultiTermAware
  (Jack Krupansky via hossman)

API Changes

* LUCENE-4391, LUCENE-4440: All methods of Lucene40Codec but
  getPostingsFormatForField are now final. To reuse functionality 
  of Lucene40, you should extend FilterCodec and delegate to Lucene40
  instead of extending Lucene40Codec.  (Adrien Grand, Shai Erea,
  Robert Muir, Uwe Schindler)

* LUCENE-4299: Added Terms.hasPositions() and Terms.hasOffsets().
  Previously you had no real way to know that a term vector field
  had positions or offsets, since this can be configured on a 
  per-field-per-document basis. (Robert Muir)

* Removed DocsAndPositionsEnum.hasPayload() and simplified the
  contract of getPayload(). It returns null if there is no payload,
  otherwise returns the current payload. You can now call it multiple
  times per position if you want. (Robert Muir)

* Removed FieldsEnum. Fields API instead implements Iterable<String>
  and exposes Iterator, so you can iterate over field names with
  for (String field : fields) instead.  (Robert Muir)

* LUCENE-4152: added IndexReader.leaves(), which lets you enumerate
  the leaf atomic reader contexts for all readers in the tree.
  (Uwe Schindler, Robert Muir)

* LUCENE-4304: removed PayloadProcessorProvider. If you want to change
  payloads (or other things) when merging indexes, its recommended
  to just use a FilterAtomicReader + IndexWriter.addIndexes. See the
  OrdinalMappingAtomicReader and TaxonomyMergeUtils in the facets
  module if you want an example of this.
  (Mike McCandless, Uwe Schindler, Shai Erera, Robert Muir)

* LUCENE-4304: Make CompositeReader.getSequentialSubReaders()
  protected. To get atomic leaves of any IndexReader use the new method
  leaves() (LUCENE-4152), which lists AtomicReaderContexts including
  the doc base of each leaf.  (Uwe Schindler, Robert Muir)

* LUCENE-4307: Renamed IndexReader.getTopReaderContext to 
  IndexReader.getContext.  (Robert Muir)

* LUCENE-4316: Deprecate Fields.getUniqueTermCount and remove it from
  AtomicReader. If you really want the unique term count across all
  fields, just sum up Terms.size() across those fields. This method
  only exists so that this statistic can be accessed for Lucene 3.x 
  segments, which don't support Terms.size().  (Uwe Schindler, Robert Muir)

* LUCENE-4321: Change CharFilter to extend Reader directly, as FilterReader
  overdelegates (read(), read(char[], int, int), skip, etc). This made it
  hard to implement CharFilters that were correct. Instead only close() is
  delegated by default: read(char[], int, int) and correct(int) are abstract
  so that its obvious which methods you should implement.  The protected 
  inner Reader is 'input' like CharFilter in the 3.x series, instead of 'in'.  
  (Dawid Weiss, Uwe Schindler, Robert Muir)

* LUCENE-3309: The expert FieldSelector API, used to load only certain
  fields in a stored document, has been replaced with the simpler
  StoredFieldVisitor API.  (Mike McCandless)

* LUCENE-4343: Made Tokenizer.setReader final. This is a setter that should
  not be overriden by subclasses: per-stream initialization should happen
  in reset().  (Robert Muir)

* LUCENE-4377: Remove IndexInput.copyBytes(IndexOutput, long). 
  Use DataOutput.copyBytes(DataInput, long) instead.
  (Mike McCandless, Robert Muir)

* LUCENE-4355: Simplify AtomicReader's sugar methods such as termDocsEnum,
  termPositionsEnum, docFreq, and totalTermFreq to only take Term as a
  parameter. If you want to do expert things such as pass a different
  Bits as liveDocs, then use the flex apis (fields(), terms(), etc) directly.
  (Mike McCandless, Robert Muir)

* LUCENE-4425: clarify documentation of StoredFieldVisitor.binaryValue
  and simplify the api to binaryField(FieldInfo, byte[]).
  (Adrien Grand, Robert Muir)

Bug Fixes

* LUCENE-4423: DocumentStoredFieldVisitor.binaryField ignored offset and
  length. (Adrien Grand)

* LUCENE-4297: BooleanScorer2 would multiply the coord() factor
  twice for conjunctions: for most users this is no problem, but
  if you had a customized Similarity that returned something other
  than 1 when overlap == maxOverlap (always the case for conjunctions),
  then the score would be incorrect.  (Pascal Chollet, Robert Muir)

* LUCENE-4298: MultiFields.getTermDocsEnum(IndexReader, Bits, String, BytesRef)
  did not work at all, it would infinitely recurse.
  (Alberto Paro via Robert Muir)


src/Server.java
/**
 * Created by frank on 17-5-26.
 */


import io.netty.bootstrap.ServerBootstrap;
import io.netty.buffer.Unpooled;
import io.netty.channel.ChannelFuture;
import io.netty.channel.ChannelInitializer;
import io.netty.channel.ChannelOption;
import io.netty.channel.EventLoopGroup;
import io.netty.channel.nio.NioEventLoopGroup;
import io.netty.channel.socket.SocketChannel;
import io.netty.channel.socket.nio.NioServerSocketChannel;
import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.ChannelInboundHandlerAdapter;
import io.netty.handler.codec.ByteToMessageDecoder;
import io.netty.util.CharsetUtil;

import java.util.List;

public class Server {

    private int port;

    public Server(int port) {
        this.port = port;
    }

    public void run() throws Exception {
        EventLoopGroup bossGroup = new NioEventLoopGroup(); // (1)
        EventLoopGroup workerGroup = new NioEventLoopGroup();
        try {
            ServerBootstrap b = new ServerBootstrap(); // (2)
            b.group(bossGroup, workerGroup)
                    .channel(NioServerSocketChannel.class) // (3)
                    .childHandler(new ChannelInitializer<SocketChannel>() { // (4)
                        @Override
                        public void initChannel(SocketChannel ch) throws Exception {
                            ch.pipeline().addLast(new MessageDecoder(), new ServerHandler());
                        }
                    })
                    .option(ChannelOption.SO_BACKLOG, 128)          // (5)
                    .childOption(ChannelOption.SO_KEEPALIVE, true); // (6)

            // Bind and start to accept incoming connections.
            ChannelFuture f = b.bind(port).sync(); // (7)

            // Wait until the server socket is closed.
            // In this example, this does not happen, but you can do that to gracefully
            // shut down your server.
            f.channel().closeFuture().sync();
        } finally {
            workerGroup.shutdownGracefully();
            bossGroup.shutdownGracefully();
        }
    }

    public static void main(String[] args) throws Exception {
        int port;
        if (args.length > 0) {
            port = Integer.parseInt(args[0]);
        } else {
            port = 8080;
        }
        System.out.println("server listening port " + port);
        new Server(port).run();
    }
}

/**
 * Handles a server-side channel.
 */
class ServerHandler extends ChannelInboundHandlerAdapter {

    @Override
    public void channelRead(ChannelHandlerContext ctx, Object msg) {
        ByteBuf in = (ByteBuf)msg;
        String query = in.toString(CharsetUtil.UTF_8);

        System.out.println("\nquery from " + ctx.channel().remoteAddress());

        /* deal with s */
        String str = SearchEngine.instance().search(query);
        // System.out.print(str);
        ByteBuf out = Unpooled.copiedBuffer(str, CharsetUtil.UTF_8);
        ctx.writeAndFlush(out);
    }

    @Override
    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
        // Close the connection when an exception is raised.
        cause.printStackTrace();
        ctx.close();
    }
}


class MessageDecoder extends ByteToMessageDecoder { // (1)
    @Override
    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) {
        String s = in.toString(io.netty.util.CharsetUtil.US_ASCII);
        int split = s.indexOf("\r\n\r\n");
        if (split == -1) {
            return;
        }

        /* 完整的请求交给上层 */
        out.add(in.readBytes(split+2));
        /* 除去空行 */
        in.readBytes(2);
    }
}

src/IndexBuilder.java
import org.apache.lucene.analysis.Analyzer;
import org.apache.lucene.document.*;
import org.apache.lucene.index.IndexWriter;
import org.apache.lucene.index.IndexWriterConfig;
import org.apache.lucene.store.Directory;
import org.apache.lucene.store.SimpleFSDirectory;
import org.apache.lucene.util.Version;
import org.wltea.analyzer.lucene.IKAnalyzer;

import java.io.*;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Collections;
import java.util.HashSet;

/**
 * Created by frank on 17-5-22.
 */
public class IndexBuilder
{
    public static void main(String[] args) throws Exception
    {
        String dataPath = "./data/meta_data/songs";
        String indexPath = "./index";

        System.out.println("building index...");
        removeDir(new File(indexPath));
        buildIndex(dataPath, indexPath);
        System.out.println("done");
    }

    private static void buildIndex(String dataPath, String indexPath) throws Exception
    {
        Analyzer analyzer = new IKAnalyzer(true);
        Directory directory = new SimpleFSDirectory(new File(indexPath));
        IndexWriterConfig config = new IndexWriterConfig(Version.LUCENE_40, analyzer);
        IndexWriter indexWriter = new IndexWriter(directory, config);

        File[] files = new File(dataPath).listFiles();

        for (File f : files) {
            BufferedReader reader =
                    new BufferedReader(
                            new InputStreamReader(
                                    new FileInputStream(f)));


            String id = reader.readLine().split(":")[1].trim();
            String song = reader.readLine().split(":")[1].trim();
            reader.readLine(); // artist_id
            String singer = reader.readLine().split("\\[|\\]")[1].trim();
            reader.readLine(); // album_id
            reader.readLine(); // album_name
            String publishTime = reader.readLine().split(":")[1].trim();
            String popularity = reader.readLine().split(":")[1].trim();

            String lrc = readLrc(id, "./data/lyrics");

            addDoc(indexWriter,
                    Integer.valueOf(id),
                    song,
                    singer,
                    Double.valueOf(publishTime),
                    Double.valueOf(popularity),
                    lrc);

            reader.close();
        }

        indexWriter.close();
        directory.close();
    }

    private static String readLrc(String id, String path)
    {
        File lrc = new File(path + "/" +id + ".lrc");

        BufferedReader reader;
        try {
            reader =
                    new BufferedReader(
                            new InputStreamReader(
                                    new FileInputStream(lrc)));
        }
        catch (Exception e) {
            return "";
        }

        char[] buffer = new char[2048];
        try {
            reader.read(buffer, 0, 2048);
        }
        catch (Exception e) {
            return "";
        }

        return String.valueOf(buffer);
    }

    private static void addDoc(IndexWriter writer,
                               int id,
                               String song,
                               String singer,
                               double publishTime,
                               double popularity,
                               String lrc) throws Exception
    {
        Document doc = new Document();
        doc.add(new IntField("id", id, IntField.TYPE_STORED));
        doc.add(new Field("song", song, TextField.TYPE_STORED));
        doc.add(new Field("singer", singer, TextField.TYPE_STORED));
        doc.add(new DoubleField("publishTime", publishTime, DoubleField.TYPE_STORED));
        doc.add(new DoubleField("popularity", popularity, DoubleField.TYPE_STORED));
        doc.add(new Field("lrc", lrc, TextField.TYPE_STORED));
        writer.addDocument(doc);
    }

    private static void removeDir(File file) throws Exception
    {
        if (file.isDirectory()) {
            for (File sub : file.listFiles()) {
                removeDir(sub);
            }
        }
        file.delete();
    }
}


src/SingerDicBuilder.java
import java.io.*;
import java.util.Collections;
import java.util.HashSet;

/**
 * Created by frank on 17-5-28.
 */
public class SingerDicBuilder
{
    public static void main(String[] args) throws Exception
    {
        String dataPath = "./data/meta_data/songs";
        System.out.println("building singer dictionary...");
        buildSingerDic(dataPath);
        System.out.println("done");
    }

    private static void buildSingerDic(String dataPath) throws Exception
    {
        File[] files = new File(dataPath).listFiles();
        HashSet<String> singers = new HashSet<>();

        for (File f : files) {
            BufferedReader reader =
                    new BufferedReader(
                            new InputStreamReader(
                                    new FileInputStream(f)));
            reader.readLine();
            reader.readLine();
            reader.readLine();
            String singer = reader.readLine().split("[\\[|\\]]")[1].trim();
            Collections.addAll(singers, singer.split("['|,]"));
            reader.close();
        }

        BufferedWriter writer =
                new BufferedWriter(
                        new OutputStreamWriter(
                                new FileOutputStream("./src/singer.dic")));
        for (String s : singers) {
            writer.write(s);
            writer.newLine();
        }
        writer.flush();
        writer.close();
    }
}


src/org/wltea/analyzer/sample/IKAnalzyerDemo.java
/**
 * IK 中文分词  版本 5.0.1
 * IK Analyzer release 5.0.1
 * 
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * 源代码由林良益(linliangyi2005@gmail.com)提供
 * 版权声明 2012，乌龙茶工作室
 * provided by Linliangyi and copyright 2012 by Oolong studio
 * 
 * 
 */
package org.wltea.analyzer.sample;

import java.io.IOException;
import java.io.StringReader;

import org.apache.lucene.analysis.Analyzer;
import org.apache.lucene.analysis.TokenStream;
import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;
import org.apache.lucene.analysis.tokenattributes.OffsetAttribute;
import org.apache.lucene.analysis.tokenattributes.TypeAttribute;
import org.wltea.analyzer.lucene.IKAnalyzer;

/**
 * 使用IKAnalyzer进行分词的演示
 * 2012-10-22
 *
 */
public class IKAnalzyerDemo {
	
	public static void main(String[] args){
		//构建IK分词器，使用smart分词模式
		Analyzer analyzer = new IKAnalyzer(true);
		
		//获取Lucene的TokenStream对象
	    TokenStream ts = null;
		try {
			ts = analyzer.tokenStream("myfield", new StringReader("这是一个中文分词的例子，你可以直接运行它！IKAnalyer can analysis english text too"));
			//获取词元位置属性
		    OffsetAttribute  offset = ts.addAttribute(OffsetAttribute.class); 
		    //获取词元文本属性
		    CharTermAttribute term = ts.addAttribute(CharTermAttribute.class);
		    //获取词元文本属性
		    TypeAttribute type = ts.addAttribute(TypeAttribute.class);
		    
		    
		    //重置TokenStream（重置StringReader）
			ts.reset();
			//迭代获取分词结果
			while (ts.incrementToken()) {
			  System.out.println(offset.startOffset() + " - " + offset.endOffset() + " : " + term.toString() + " | " + type.type());
			}
			//关闭TokenStream（关闭StringReader）
			ts.end();   // Perform end-of-stream operations, e.g. set the final offset.

		} catch (IOException e) {
			e.printStackTrace();
		} finally {
			//释放TokenStream的所有资源
			if(ts != null){
		      try {
				ts.close();
		      } catch (IOException e) {
				e.printStackTrace();
		      }
			}
	    }
		
	}

}


src/SearchEngine.java
/**
 * Created by frank on 17-5-22.
 */
import com.sun.javaws.exceptions.InvalidArgumentException;
import javafx.util.Pair;
import org.apache.lucene.analysis.*;
import org.apache.lucene.document.*;
import org.apache.lucene.index.*;
import org.apache.lucene.queryparser.classic.QueryParser;
import org.apache.lucene.search.*;
import org.apache.lucene.search.highlight.*;
import org.apache.lucene.store.Directory;
import org.apache.lucene.store.SimpleFSDirectory;
import org.apache.lucene.util.Version;
import org.wltea.analyzer.lucene.IKAnalyzer;

import java.io.*;
import java.util.*;

public class SearchEngine
{
    private final String FieldId = "id";
    private final String FieldSong = "song";
    private final String FieldSinger = "singer";
    private final String FieldLrc = "lrc";
    private final String FieldAll = "all";
    private final String FieldPublishTime = "publishTime";
    private final String FieldPopularity = "popularity";
    private final String ActionUpdate = "update";

    private final Double lrcThreshold = 3.0;

    class MyDoc
    {
        public MyDoc(ScoreDoc sd, Query q, String fieldName, String fieldValue) throws Exception
        {
            score = sd.score;
            doc = indexSearcher.doc(sd.doc);

            setField(fieldName, fieldValue, q);

            id = Integer.valueOf(doc.get(FieldId));
            song = doc.get(FieldSong);
            singer = doc.get(FieldSinger);
            lrc = doc.get(FieldLrc);
            publishTime = Double.valueOf(doc.get(FieldPublishTime));
            popularity = Double.valueOf(doc.get(FieldPopularity));
        }

        public MyDoc(MyDoc lhs, MyDoc rhs)
        {
            assert (lhs.id == rhs.id);

            score = lhs.score * rhs.score;
            doc = lhs.doc;

            setField(lhs);
            setField(rhs);

            id = lhs.id;
            song = lhs.song;
            singer = lhs.singer;
            lrc = lhs.lrc;
            publishTime = lhs.publishTime;
            popularity = lhs.popularity;
        }

        public MyDoc(MyDoc d, double newSore)
        {
            score = newSore;
            doc = d.doc;

            setField(d);

            id = d.id;
            song = d.song;
            singer = d.singer;
            lrc = d.lrc;
            publishTime = d.publishTime;
            popularity = d.popularity;
        }

        public void setField(String fieldName, String fieldValue, Query query)
        {
            switch (fieldName) {
                case FieldSong:
                    songQueried = true;
                    songQueryValue = fieldValue;
                    songQuery = query;
                    break;

                case FieldSinger:
                    singerQueried = true;
                    singerQueryValue = fieldValue;
                    singerQuery = query;
                    break;

                case FieldLrc:
                    lrcQueried = true;
                    lrcQueryValue = fieldValue;
                    lrcQuery = query;
            }
        }

        public void setField(MyDoc doc)
        {
            if (doc.songQueried) {
                songQueried = true;
                songQueryValue = doc.songQueryValue;
                songQuery = doc.songQuery;
            }
            if (doc.singerQueried) {
                singerQueried = true;
                singerQueryValue = doc.singerQueryValue;
                singerQuery = doc.singerQuery;
            }
            if (doc.lrcQueried) {
                lrcQueried = true;
                lrcQueryValue = doc.lrcQueryValue;
                lrcQuery = doc.lrcQuery;
            }
        }

        @Override
        public String toString()
        {
            String hSong = null;
            String hSinger = null;
            String hLrc = null;

            if (songQueried) {
                hSong = highlightFiledValue(songQuery, FieldSong, song);
            }
            if (singerQueried) {
                hSinger = highlightFiledValue(singerQuery, FieldSinger, singer);
            }
            if (lrcQueried) {
                hLrc = highlightFiledValue(lrcQuery, FieldLrc, lrc);
            }

            String ret = String.format("id: %d\r\n", id);
            ret += String.format("song: %s\r\n", hSong != null ? hSong : song);
            ret += String.format("singer: %s\r\n", hSinger != null ? hSinger : singer);
            if (hLrc != null) {
                ret += String.format("lrc: %s\r\n", hLrc);
            }

            return ret;
        }

        public double score;
        public Document doc;

        public boolean songQueried = false;
        public String songQueryValue;
        public Query songQuery;

        public boolean singerQueried = false;
        public String singerQueryValue;
        public Query singerQuery;

        public boolean lrcQueried = false;
        public String lrcQueryValue;
        public Query lrcQuery;

        public int id;
        public String song;
        public String singer;
        public String lrc;
        public double publishTime;
        public double popularity;
    }

    private static SearchEngine singleton;

    public static SearchEngine instance()
    {
        try {
            if (singleton == null) {
                singleton = new SearchEngine();
            }
        }catch (Exception e) {
            e.printStackTrace();
        }

        return singleton;
    }

    private SearchEngine() throws Exception
    {
        analyzer = new IKAnalyzer(true);
        directory = new SimpleFSDirectory(new File("./index"));
        directoryReader = DirectoryReader.open(directory);
        indexSearcher = new IndexSearcher(directoryReader);
        clickTimes = new HashMap<>();
        singers = new HashSet<>();

        BufferedReader reader =
                new BufferedReader(
                        new InputStreamReader(
                                new FileInputStream("src/singer.dic")));

        try {
            while (true) {
                String s = reader.readLine();
                if (!s.isEmpty()) {
                    singers.add(s);
                }
            }
        }
        catch (Exception e) {
            System.out.println(singers.size() + " singers");
        }

        System.out.println(directoryReader.numDocs() + " docs in search engine");
    }

    public void update(int songId, int times)
    {
        System.out.printf("update %d %d times\n", songId, times);
        clickTimes.put(songId, times);
    }

    public String search(String input)
    {
        Vector<Pair<String, String>> query = parseQuery(input);
        MyDoc[] docs = null;

        try {
            for (Pair<String, String> p : query) {
                String key = p.getKey();
                String value = p.getValue();
                MyDoc[] newDocs = null;
                switch (key) {

                    case ActionUpdate:
                        String[] s = value.split("[ |\t]");
                        if (s.length != 2) {
                            throw new InvalidArgumentException(null);
                        }
                        update(Integer.valueOf(s[0]), Integer.valueOf(s[1]));
                        break;

                    case FieldId:
                    case FieldSong:
                    case FieldSinger:
                    case FieldLrc:
                        newDocs = searchField(key, value);
                        break;

                    case FieldAll:
                        if (singers.contains(value)) {
                            // 优先从词典中匹配歌手名
                            newDocs = searchField(FieldSinger, value);
                            System.out.printf("hits %d\n", newDocs.length);
                        }
                        else {
                            // 歌名
                            newDocs = searchField(FieldSong, value);
                            sortDocs(newDocs);
                            System.out.printf("hits %d\n", newDocs.length);
                            if (newDocs.length > 0) {
                                System.out.printf("song score %f \n", newDocs[0].score);
                            }
                            if (newDocs.length == 0 || newDocs[0].score < lrcThreshold) {
                                // 歌词优先级最低
                                newDocs = searchField(FieldLrc, value);
                                sortDocs(newDocs);
                                System.out.printf("hits %d\n", newDocs.length);
                                if (newDocs.length > 0) {
                                    System.out.printf("lrc score %f \n", newDocs[0].score);
                                }
                            }
                        }
                        break;

                    default:
                        throw new InvalidArgumentException(null);
                }

                if (docs == null) {
                    docs = newDocs;
                } else {
                    docs = intersectDocs(docs, newDocs);
                }
            }
        }
        catch (Exception e) {
            e.printStackTrace();
        }

        if (docs == null) {
            docs = new MyDoc[0];
        }

        sortDocs(docs);
        StringBuilder buf = new StringBuilder(128);
        buf.append(String.format("hits: %d\r\n\r\n", docs.length));
        for (MyDoc d : docs) {
            buf.append(d.toString());
            buf.append("\r\n");
        }

        return buf.toString();
    }

    private Vector<Pair<String, String>> parseQuery(String input)
    {
        BufferedReader reader =
                new BufferedReader(new StringReader(input));
        Vector<Pair<String, String>> ret = new Vector<>(1);

        try {
            while (true) {
                String line = reader.readLine();
                if (line == null) {
                    break;
                }

                String[] pair = line.split(":");
                if (pair.length != 2) {
                    throw new IOException("invalid query");
                }

                pair[0] = pair[0].trim();
                pair[1] = pair[1].trim();

                if (pair[0].equalsIgnoreCase(FieldSong)) {
                    pair[0] = FieldSong;
                }
                else if (pair[0].equalsIgnoreCase(FieldSinger)) {
                    pair[0] = FieldSinger;
                }
                else if (pair[0].equalsIgnoreCase(FieldLrc)) {
                    pair[0] = FieldLrc;
                }
                else if (pair[0].equalsIgnoreCase(FieldAll)) {
                    pair[0] = FieldAll;
                }
                else if (pair[0].equalsIgnoreCase(ActionUpdate)) {
                    pair[0] = ActionUpdate;
                }

                else {
                    throw new IOException("invalid query");
                }
                ret.add(new Pair<>(pair[0], pair[1]));
            }
        }
        catch (Exception e) {
            e.printStackTrace();
        }

        return ret;
    }

    private MyDoc[] searchField(String name, String value) throws Exception
    {
        QueryParser parser = new QueryParser(Version.LUCENE_40, name, analyzer);
        Query query = parser.parse(value);
        ScoreDoc[] sdocs = indexSearcher.search(query, 100).scoreDocs;
        MyDoc[] mdocs = new MyDoc[sdocs.length];


        for (int i = 0; i < mdocs.length; ++i) {
            mdocs[i] = new MyDoc(sdocs[i], query, name, value);

            int id = mdocs[i].id;
            int times = clickTimes.getOrDefault(id, 0);

            mdocs[i].score = sdocs[i].score * (1 + Math.log(1 + times));
        }

        System.out.println(query);

        return mdocs;
    }

    private String highlightFiledValue(Query q, String fieldName, String text)
    {
        SimpleHTMLFormatter simpleHTMLFormatter =
                new SimpleHTMLFormatter(
                        "*", "*");

        Highlighter highlighter =
                new Highlighter(simpleHTMLFormatter, new QueryScorer(q));

        highlighter.setTextFragmenter(new SimpleFragmenter(20));

        try {
            return highlighter.getBestFragment(analyzer, fieldName, text);
        }
        catch (Exception e) {
            e.printStackTrace();
        }
        return "";
    }

    private void sortDocs(MyDoc[] docs)
    {
        /* 根据score, popularity, publishTime排序 */
        class MyComparator implements Comparator<MyDoc>
        {
            public int compare(MyDoc lhs, MyDoc rhs)
            {
                if (lhs.score != rhs.score) {
                    return lhs.score < rhs.score ? 1 : -1;
                }

                if (lhs.popularity != rhs.popularity) {
                    return lhs.popularity < rhs.popularity ? -1 : 1;
                }

                return lhs.publishTime > rhs.publishTime ? -1 : 1;
            }
        }
        Arrays.sort(docs, new MyComparator());
    }

    private int nToken(String input) throws Exception
    {
        int tokens = 0;
        TokenStream ts = analyzer.tokenStream("myfield", new StringReader(input));

        //重置TokenStream（重置StringReader）
        ts.reset();

        //迭代获取分词结果
        while (ts.incrementToken()) {
            ++tokens;
        }

        //关闭TokenStream（关闭StringReader）
        ts.end();

        return tokens;
    }

    private MyDoc[] intersectDocs(MyDoc[] a, MyDoc[] b)
    {
        Vector<MyDoc> vec = new Vector<>();
        for (MyDoc d1 : a) {
            for (MyDoc d2 : b) {
                if (d1.id == d2.id) {
                    vec.add(new MyDoc(d1, d2));
                }
            }
        }

        MyDoc[] arr = new MyDoc[vec.size()];
        return vec.toArray(arr);
    }

    private MyDoc[] unionDocs(MyDoc[] a, MyDoc[] b)
    {
        Vector<MyDoc> vec = new Vector<>();
        Collections.addAll(vec, b);
        for (MyDoc d1 : a) {
            boolean found = false;
            for (MyDoc d2 : b) {
                if (d1.id == d2.id) {
                    found = true;
                    break;
                }
            }
            if (!found) {
                vec.add(d1);
            }
        }
        MyDoc[] arr = new MyDoc[vec.size()];
        return vec.toArray(arr);
    }

    private HashSet<String> singers;
    private Analyzer analyzer;
    private Directory directory;
    private DirectoryReader directoryReader;
    private IndexSearcher indexSearcher;
    private HashMap<Integer, Integer> clickTimes;

    public static void main(String[] args) throws Exception
    {
        SearchEngine se = new SearchEngine();
        BufferedReader reader =
                new BufferedReader(
                        new InputStreamReader(System.in));

        String input = "singer: 周杰伦\r\nsong: 搁浅\r\n";
        se.search(input);

        while (true) {

            System.out.print("search: ");
            input = reader.readLine();

            se.search(input);

            //System.out.print("click: ");
            //int click = Integer.valueOf(reader.readLine());
            //se.update(click);
        }
    }
}